From ffedc5ab6467e285cfe225cebe6dbf368f1bed4d Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 20 Jul 2023 20:22:31 +0800
Subject: [PATCH 08/53] [Huawei] mm/sharepool: Delete file element in
 sp_alloc_context

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

It is used to decide whether we should fallback while failing to
allocate hugepage. We can use ac->sp_flags to do the same thing.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 27 +++++++++++----------------
 1 file changed, 11 insertions(+), 16 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 8a26a7f1fbfe..ca6178262d1f 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -2149,7 +2149,6 @@ static unsigned long sp_mmap(struct mm_struct *mm, struct file *file,
 
 struct sp_alloc_context {
 	struct sp_group *spg;
-	struct file *file;
 	unsigned long size;
 	unsigned long size_aligned;
 	unsigned long sp_flags;
@@ -2211,13 +2210,10 @@ static int sp_alloc_prepare(unsigned long size, unsigned long sp_flags,
 	}
 
 	ac->type = SPA_TYPE_ALLOC;
-	if (sp_flags & SP_HUGEPAGE) {
-		ac->file = spg->file_hugetlb;
+	if (sp_flags & SP_HUGEPAGE)
 		ac->size_aligned = ALIGN(size, PMD_SIZE);
-	} else {
-		ac->file = spg->file;
+	else
 		ac->size_aligned = ALIGN(size, PAGE_SIZE);
-	}
 
 	ac->spg = spg;
 	ac->size = size;
@@ -2229,20 +2225,19 @@ static int sp_alloc_prepare(unsigned long size, unsigned long sp_flags,
 
 static void sp_alloc_fallback(struct sp_area *spa, struct sp_alloc_context *ac)
 {
-	if (ac->file == ac->spg->file) {
+	/*
+	 * If hugepage allocation fails, this will transfer to normal page
+	 * and try again. (only if SP_HUGEPAGE_ONLY is not flagged
+	 */
+	if (!(ac->sp_flags & SP_HUGEPAGE) || (ac->sp_flags & SP_HUGEPAGE_ONLY)) {
 		ac->state = ALLOC_NOMEM;
 		return;
 	}
 
-	if (!(ac->sp_flags & SP_HUGEPAGE_ONLY)) {
-		ac->file = ac->spg->file;
-		ac->size_aligned = ALIGN(ac->size, PAGE_SIZE);
-		ac->sp_flags &= ~SP_HUGEPAGE;
-		ac->state = ALLOC_RETRY;
-		__sp_area_drop(spa);
-		return;
-	}
-	ac->state = ALLOC_NOMEM;
+	ac->size_aligned = ALIGN(ac->size, PAGE_SIZE);
+	ac->sp_flags &= ~SP_HUGEPAGE;
+	ac->state = ALLOC_RETRY;
+	__sp_area_drop(spa);
 }
 
 static long sp_mbind(struct mm_struct *mm, unsigned long start, unsigned long len,
-- 
2.17.1

